README.md

![version:1.1.0](https://img.shields.io/badge/version-1.1.0-f53.svg)
![license:BSD3](https://img.shields.io/badge/license-BSD3-0d0.svg)

# multi_pull

"multi pull" is an extension of RefreshIndicator!

I'd love it if you gave me a star!🌟

## No words are needed at first. Just look at it.

<img width=250 src="https://user-images.githubusercontent.com/61507019/119260538-32a26680-bc0e-11eb-94ac-7c341a00aa79.gif">


### one of example customized multi_pull

<img width=250 src="https://user-images.githubusercontent.com/61507019/136678547-27f96ef3-c3b3-4872-90f9-a3a693b72701.png">


## What is MultiPull?

The RefreshIndicator is mainly used for scrolling and reloading, while MultiPull is an extension of the RefreshIndicator that allows users to select an action based on the widget that appears when they scroll. I have placed actions in the appBar, but I want users to be able to access them more easily. If you have an action in the appBar, but want to make it easier to access, you can place the same action in the MultiPull, and users will be able to reach it by just scrolling!

So long story short, let's see it in action!


## How to install

```pubspec.yml
dependencies:
  multi_pull: 1.1.0
```

## Getting Started

The usage of MultiPull is similar to RefreshIndicator. It works by having a Scrollable Widget like a ListView as a child. Then you place the widget in actionWidgets, wrapped in an ActionWidget that is displayed as a choice when pulled. Don't forget to set the action or onRefresh to what will happen when the choice is made.

```dart:main.dart
import 'package:multi_pull/multi_pull.dart';

~~~~
~~~~
    Scaffold(
      appBar: AppBar(
        title: Text("Second Page"),
      ),
      body: MultiPull(
        circleMoveDuration: Duration(milliseconds: 400),
        circleMoveCurve: Curves.easeInOut,
        circleIndicator: DefaultCircle(
          circleColor: Colors.grey,
          circleOpacity: 0.2,
        ),
        pullIndicators: [
          DefaultPullIndicator(
            icon: Icon(Icons.arrow_back_ios_outlined),
            label: Text("back"),
            onPull: () => Navigator.pop(context),
          ),
          DefaultPullIndicator(
            icon: Icon(Icons.refresh_rounded),
            onPull: () async => await Future.delayed(Duration(seconds: 2)),
          ),
          DefaultPullIndicator(
            icon: Icon(Icons.backspace_outlined),
            label: Text("clear", style: TextStyle(color: Colors.redAccent)),
            onPull: () {
              _firstTextController.clear();
              _secondTextController.clear();
              _thirdTextController.clear();
            },
          ),
        ],
        child: ListView(
          physics: AlwaysScrollableScrollPhysics(),
          children: [
            TextField(
              controller: _firstTextController,
            ),
            TextField(
              controller: _secondTextController,
            ),
            TextField(
              controller: _thirdTextController,
            ),
          ],
        ),
      ),
    );
```

The difference between action and onRefresh is simple: if you want to do asynchronous processing, just use onRefresh. The difference between action and onRefresh is simple: if you want to do asynchronous processing, just use onRefresh. no, technically, action can also do asynchronous processing, but onRefresh displays a RefreshIndicator to indicate that it is doing asynchronous processing.


CHANGELOG.md
## [1.1.0]

- enable customize PullIndicator and CircleIndicator.

**breaking change**
- rename refresh indicator colors property.


## [1.0.0+1]

**breaking changes**
- rename ActionWidget to PullIndicator
  - then rename MultiPull property 'actionWidgets' to 'pullIndicators' too.
- abolish PullIndicator(ActionWidget)'s property: action and onRefresh.
  - then added new property onPull on MultiPull.
- label property of PullIndicator(ActionWidget) change type from String? to Widget?.
- there was some gap between user circle and pullIndicators. so fixed the misalignment.

## [0.3.2] -

- fix ActionWidget alignment when no label.
- breaking change: change ActionWidget label type from String to Widget.

## [0.3.1] -

- fix error: colortween is not subtype of type animated<color>

## [0.2.1] -

- update document

## [0.2.0] -

- fix bug: when build with ios device, warn ".xconfig file not exist". -> cd ios && pod install

- new 4 variables about "forcus circle"!!
  - circleOpacity: it just circle opacity
  - circleColor: it just circle color.
  - circleMoveDuration: circle moving animatioin duration
  - circleMoveCurve: circle moving animation curve


## [0.1.0] -

* initial release

example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:multi_pull/multi_pull.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            ElevatedButton(
              child: Text('simple pull page'),
              onPressed: () => Navigator.push(
                  context,
                  MaterialPageRoute(
                    builder: (context) => SimplePage(),
                  )),
            ),
            ElevatedButton(
              child: Text('label pull page'),
              onPressed: () => Navigator.push(
                  context,
                  MaterialPageRoute(
                    builder: (context) => LabelPage(),
                  )),
            ),
          ],
        ),
      ),
    );
  }
}

class SimplePage extends StatefulWidget {
  SimplePage({Key? key}) : super(key: key);

  @override
  SimplePageState createState() => SimplePageState();
}

class SimplePageState extends State<SimplePage> {
  SimplePageState();

  final _firstTextController = TextEditingController();
  final _secondTextController = TextEditingController();
  final _thirdTextController = TextEditingController();

  @override
  void initState() {
    super.initState();
    _firstTextController.text = "Hello";
    _secondTextController.text = "multi pull actions";
    _thirdTextController.text = "airy-swift";
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Second Page"),
      ),
      body: MultiPull(
        circleMoveDuration: Duration(milliseconds: 400),
        circleMoveCurve: Curves.easeInOut,
        circleIndicator: DefaultCircle(
          circleColor: Colors.grey,
          circleOpacity: 0.2,
        ),
        pullIndicators: [
          DefaultPullIndicator(
            icon: Icon(Icons.arrow_back_ios_outlined),
            label: Text("back"),
            onPull: () => Navigator.pop(context),
          ),
          DefaultPullIndicator(
            icon: Icon(Icons.refresh_rounded),
            onPull: () async => await Future.delayed(Duration(seconds: 2)),
          ),
          DefaultPullIndicator(
            icon: Icon(Icons.backspace_outlined),
            label: Text("clear", style: TextStyle(color: Colors.redAccent)),
            onPull: () {
              _firstTextController.clear();
              _secondTextController.clear();
              _thirdTextController.clear();
            },
          ),
        ],
        child: ListView(
          physics: AlwaysScrollableScrollPhysics(),
          children: [
            TextField(
              controller: _firstTextController,
            ),
            TextField(
              controller: _secondTextController,
            ),
            TextField(
              controller: _thirdTextController,
            ),
          ],
        ),
      ),
    );
  }
}

class LabelPage extends StatelessWidget {
  const LabelPage();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: MultiPull(
        circleMoveDuration: Duration(milliseconds: 400),
        circleMoveCurve: Curves.easeInOut,
        circleIndicator: ArrowIndicator(),
        pullIndicators: [
          LabelPullIndicator(
            label: "Back",
            onPull: () => Navigator.pop(context),
          ),
          LabelPullIndicator(
            label: "Refresh",
            onPull: () async => await Future.delayed(Duration(seconds: 2)),
          ),

        ],
        child: ListView(
          physics: AlwaysScrollableScrollPhysics(),
          children: [
            Center(child: Text("")),
          ],
        ),
      ),
    );
  }
}